pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,458 @@
+	// in the queue.
+	eventTypeDropped
+)
+
+var eventTypeStrings = map[eventType]string{
+	eventTypeRetry:   "eventTypeRetry",
+	eventTypeFlushed: "eventTypeFlushed",
+	eventTypeSent:    "eventTypeSent",
+	eventTypeFailed:  "eventTypeFailed",
+	eventTypeDropped: "eventTypeDropped",
+}
+
+// String implements fmt.Stringer.
+func (t eventType) String() string { return eventTypeStrings[t] }
+
+// eventData represents information about a sender event. Not all fields apply
+// to all events.
+type eventData struct {
+	// host specifies the host which the sender is sending to.
+	host string
+	// bytes represents the number of bytes affected by this event. It is
+	// not known for eventTypeFlushed.
+	bytes int
+	// count specfies the number of payloads that this events refers to.
+	count int
+	// duration specifies the time it took to complete this event. It
+	// is set for eventType{Flushed,Sent,Retry,Failed}.
+	duration time.Duration
+	// err specifies the error that may have occurred on events like
+	// eventTypeRetry and eventTypeFailed.
+	err error
+	// connectionFill specifies the percentage of allowed connections used.
+	// At 100% (1.0) the writer will become blocking. It is reported for events
+	// eventType{Sent,Retry,Failed}.
+	connectionFill float64
+}
+
+// maxQueueSize specifies the maximum allowed queue size. If it is surpassed, older
+// items are dropped to make room for new ones.
+var maxQueueSize = 64 * 1024 * 1024 // 64MB; replaced in tests
+
+// sender is responsible for sending payloads to a given URL. It uses a size-limited
+// retry queue with a backoff mechanism in case of retriable errors.
+type sender struct {
+	cfg  *senderConfig
+	host string
+
+	climit chan struct{} // acts as a semaphore for limiting concurrent connections
+
+	mu        sync.Mutex  // guards below fields
+	list      *list.List  // send queue
+	size      int         // size of send queue
+	scheduled bool        // reports if a flush is scheduled
+	attempt   int         // retry attempt coming up next
+	timer     *time.Timer // timed flush (triggered by backoff)
+}
+
+// senderConfig specifies the configuration for the sender.
+type senderConfig struct {
+	// client specifies the HTTP client to use when sending requests.
+	client *http.Client
+	// url specifies the URL to send requests too.
+	url *url.URL
+	// apiKey specifies the Datadog API key to use.
+	apiKey string
+	// maxConns specifies the maximum number of allowed concurrent ougoing
+	// connections.
+	maxConns int
+	// recorder specifies the eventRecorder to use when reporting events occurring
+	// in the sender.
+	recorder eventRecorder
+}
+
+// newSender creates a new sender, which uses the given HTTP client, URL and API key
+// to make requests using incoming payloads. The limit of outgoing concurrent connections
+// is defined by climit.
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
```

Comment:
> * "How large should the channel be?"

We're already enforcing a size on the list though, so that would be the channel size? 

> * Draining the channel, as well as removing items from it will result in code that looks like below and would again be very unnatural

Back to my other comments about the whole flushing approach, I'm not sure draining the whole channel is necessary as I don't think flushing a group of pending payloads makes sense since they can't be batched. So all we should be doing is reading them off of the channel one-at-a-time and sending them off asynchronously. 

> Lastly, the previous implementation [also used `list.List` for the retry queue](https://github.com/DataDog/datadog-agent/blob/a08cd3555c92780bcd6b580799e1e179aaec3c2d/pkg/trace/writer/sender.go#L81), which is a natural choice for queues and can grow dynamically without it being pre-allocated. Much easier to reason about.

That probably could have been a channel as well. But as I mentioned already we might not even need a separate retry list/channel as we could just throw the retry payloads back onto the same channel.